### PR TITLE
feat: Dependent Resources for External Resources

### PR DESCRIPTION
**Pull request by csviri**
WIP

### CODE REVIEW

**Review by metacosm — CHANGES_REQUESTED** [review]

**Comment by metacosm** [comment]
I made small changes that I think improve readability a little. If we can sort out the supply thing, the return values either not being used or returning `null` vs. `Optional` and the `del` method (which I think should stay `delete), we'll be good to merge.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java [hunk 1]
```diff
@@ -4,9 +4,10 @@
 import io.javaoperatorsdk.operator.api.reconciler.Context;
 
 @SuppressWarnings("rawtypes")
+@FunctionalInterface
 public interface Deleter<P extends HasMetadata> {
   Deleter NOOP = (primary, context) -> {
   };
 
-  void delete(P primary, Context context);
+  void del(P primary, Context context);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java:12]: The problem is that is the same method name and params as with the `DependentResource`. So if implemented there is a collision, and basically for example here:
https://github.com/java-operator-sdk/java-operator-sdk/blob/af5b2e068905b5c8c24607a005c142cb952d29cb/operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/AbstractSimpleDependentResource.java#L48-L51

If sombody wnat to implement `Deleter` would override the parent logic for delete.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java [hunk 2]
```diff
@@ -62,10 +63,73 @@ public void reconcile(P primary, Context context) {
...
+    } catch (RuntimeException e) {
+      if (isEventSourceARecentOperationEventFilter()) {
+        eventSourceAsRecentOperationEventFilter()
+            .cleanupOnCreateOrUpdateEventFiltering(resourceID, created == null ? desired : created);
+      }
+      throw e;
+    }
+  }
+
+  protected R handleUpdate(R actual, R desired, P primary, Context context) {
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java]: removed it.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Creator.java [hunk 3]
```diff
@@ -4,9 +4,9 @@
...
 
 @SuppressWarnings("rawtypes")
+@FunctionalInterface
 public interface Creator<R, P extends HasMetadata> {
-  Creator NOOP = (desired, primary, context) -> {
-  };
+  Creator NOOP = (desired, primary, context) -> null;
 
-  void create(R desired, P primary, Context context);
+  R create(R desired, P primary, Context context);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Creator.java:11]: The problem is that theretically could be just in NOOP, what is a marker interface (maybe document it this?). If a resource is marked as a creator, should be able to create the resource, if not I would except to rather throw an exception

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java [hunk 4]
```diff
@@ -62,10 +63,73 @@ public void reconcile(P primary, Context context) {
...
+        eventSourceAsRecentOperationEventFilter().prepareForCreateOrUpdateEventFiltering(resourceID,
+            desired);
+      }
+      created = creator.create(desired, primary, context);
+      if (isEventSourceARecentOperationCacheFiller()) {
+        eventSourceAsRecentOperationCacheFiller().handleRecentResourceCreate(resourceID, created);
+      }
+      return created;
+    } catch (RuntimeException e) {
+      if (isEventSourceARecentOperationEventFilter()) {
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java]: This is currently done by `handleRecentCreateOrUpdate` what can be confusing I agree, will check and change it

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/AbstractSimpleDependentResource.java [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+    initMatcher();
+  }
+
+  @Override
+  public Optional<R> getResource(HasMetadata primaryResource) {
+    return cache.get(ResourceID.fromResource(primaryResource));
+  }
+
+  /** Actually read the resource from the target API */
+  public abstract Optional<R> supplyResource(HasMetadata primaryResource);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/AbstractSimpleDependentResource.java:38]: I like fetch better too :), just did not remeber that, will change it

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/AbstractCachingDependentResource.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import io.javaoperatorsdk.operator.api.reconciler.dependent.EventSourceProvider;
+import io.javaoperatorsdk.operator.processing.event.ExternalResourceCachingEventSource;
+import io.javaoperatorsdk.operator.processing.event.source.EventSource;
+
+public abstract class AbstractCachingDependentResource<R, P extends HasMetadata>
+    extends AbstractDependentResource<R, P> implements EventSourceProvider<P> {
+
+  protected ExternalResourceCachingEventSource<R, P> eventSource;
+
+  public Optional<R> getSupplierResource(P primaryResource) {
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/AbstractCachingDependentResource.java:17]: it's fetching, here the `getResource()` gets the resource from the cache. Will chanage the name

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/PerResourcePollingDependentResource.java [hunk 7]
```diff
@@ -0,0 +1,40 @@
+package io.javaoperatorsdk.operator.processing.dependent.external;
+
+import io.fabric8.kubernetes.api.model.HasMetadata;
+import io.javaoperatorsdk.operator.api.reconciler.EventSourceContext;
+import io.javaoperatorsdk.operator.processing.event.source.EventSource;
+import io.javaoperatorsdk.operator.processing.event.source.polling.PerResourcePollingEventSource;
+
+import static io.javaoperatorsdk.operator.processing.dependent.external.PollingDependentResource.DEFAULT_POLLING_PERIOD;
+
+public abstract class PerResourcePollingDependentResource<R, P extends HasMetadata>
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/dependent/external/PerResourcePollingDependentResource.java:10]: The PerResource polls the supplier for every resource separatelly, the Poller just periodically polls an api regardless of the resources in the caches of the ControllerResourceEventSource

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/polling/PerResourcePollingEventSource.java [hunk 8]
```diff
@@ -151,8 +152,8 @@ public void run() {
     }
   }
 
-  public interface ResourceSupplier<T, R> {
-    Optional<T> getResource(R resource);
+  public interface ResourceSupplier<R, P> {
+    Optional<R> getSupplierResource(P resource);
```
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/processing/event/source/polling/PerResourcePollingEventSource.java:156]: changed it, to resourceFetcher, also the name, good point.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java:12]: That's partly why I think it's a bad idea to put `delete` on `DependentResource` especially if we don't force people to implement the method by providing a default empty implementation. If people need to support the delete operation in their `DependentResource`, they should just implement `Deleter`. I think that would be cleaner. 
Another solution would be to name the method on `DependentResource` `cleanup`, which would mirror `Reconciler` and then we could keep a "real" method name for the `Deleter` interface.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Creator.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Creator.java:8]: I push a fix I think will address this issue. Let's see how the tests go.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java]: Checked this, the problem is that it needs to be in the same `synchonized` block as it is now. Will check once more, but in a separate PR, would rather stick with this now. (when changed the tests started to fail)

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java:8]: That would mix the traits and the high level api. So not sure that is nicer. But don't have strong opinion on this.

I would go with `reconcile` now. Will change it that way, maybe we can try to take a look on extraction of delete in a separate PR. If you agree.

[the remark this answers is not in the document]
Comment (csviri) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/Deleter.java]: changed to cleanup in `DependetResource`.

File: operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (metacosm) [reply operator-framework-core/src/main/java/io/javaoperatorsdk/operator/api/reconciler/dependent/AbstractDependentResource.java]: Can you create an issue to track this, please? Let's merge this!

**Review by metacosm — APPROVED** [review]

**Pull request merged**